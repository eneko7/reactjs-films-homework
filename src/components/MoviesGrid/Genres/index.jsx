import Genres from './GenresContainer';

export default Genres;
