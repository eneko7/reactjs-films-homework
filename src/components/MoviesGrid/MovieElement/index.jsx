import MovieElement from './MovieElementContainer';

export default MovieElement;
