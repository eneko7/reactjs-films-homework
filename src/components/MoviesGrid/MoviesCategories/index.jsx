import MoviesCategories from './MoviesCategoriesContainer';

export default MoviesCategories;
