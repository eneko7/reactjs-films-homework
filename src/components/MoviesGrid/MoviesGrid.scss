.moviesGrid {
  width: 100%;
  position: relative;
  z-index: 4;
  background-color: #f5f5f5;
  padding-top: 50px;

  .moviesGrid_categories {
    margin: 70px 7%;
    margin-top: 0;
    margin-bottom: 30px;
    display: flex;
    justify-content: flex-start;
  }

  .moviesGrid_wrapper {
    .moviesGrid_wrapper_MovieElement_ul {
      margin: 0 7%;
      display: flex;
      justify-content: space-between;
      flex-wrap: wrap;
      padding: 0;
      list-style-type: none;

      .moviesGrid_wrapper_MovieElement_ul_item {
        width: 23.3%;
        height: 500px;
        margin-bottom: 60px;
        border-radius: 6px;
      }
    }

    .moviesGrid_wrapper_loading {
      display: flex;
      justify-content: space-between;
      width: 70px;
      margin: 20px auto 5px auto;

      .moviesGrid_wrapper_loading {
        width: 13px;
        height: 13px;
        border-radius: 50%;
        background-color: #0aaee4;
        animation-name: loading;
        animation-duration: 1s;
        animation-iteration-count: infinite;
      }

      .moviesGrid_wrapper_loading:nth-child(2) {
        animation-delay: 0.2s;
      }

      .moviesGrid_wrapper_loading:last-child {
        animation-delay: 0.4s;
      }
    }

    .moviesGrid_wrapper_loading_text {
      display: block;
      margin: 5px auto 0 auto;
      padding-bottom: 50px;
      font-size: 20px;
      color: #969797;
      text-align: center;
      font-family: Helvetica, Arial, Verdana, Tahoma, sans-serif;
    }

    @keyframes loading {
      30% { transform: scale(1.3, 1.3); }
      70% { transform: scale(0.9, 0.9); }
    }
  }
}

.error_search {
  width: 100%;
  text-align: center;
  color: #0aaee4;
  font-size: 30px;
  margin-bottom: 50px;
}

@media (max-width: 1440px) {
  .moviesGrid {
    .moviesGrid_wrapper {
      .moviesGrid_wrapper_MovieElement_ul {
        .moviesGrid_wrapper_MovieElement_ul_item {
          width: 31.6%;
        }
      }
    }
  }
}

@media (max-width: 1200px) {
  .moviesGrid {
    .moviesGrid_wrapper {
      .moviesGrid_wrapper_MovieElement_ul {
        .moviesGrid_wrapper_MovieElement_ul_item {
          width: 48.3%;
        }
      }
    }
  }
}

@media (max-width: 766px) {
  .moviesGrid {
    .moviesGrid_wrapper {
      .moviesGrid_wrapper_MovieElement_ul {
        justify-content: space-between;

        .moviesGrid_wrapper_MovieElement_ul_item {
          width: 100%;
          height: 600px;
        }
      }
    }
  }
}

@media (min-width: 767px) and (max-width: 769px) {
  .moviesGrid {
    .moviesGrid_wrapper {
      .moviesGrid_wrapper_MovieElement_ul {
        justify-content: space-between;

        .moviesGrid_wrapper_MovieElement_ul_item {
          width: 48%;
          height: 600px;
        }
      }
    }
  }
}
