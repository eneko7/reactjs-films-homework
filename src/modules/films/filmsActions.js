/* eslint-disable max-len */
import axios from 'axios';
import { receiveMainFilmInfo } from '../film/filmActions';
import {
  urlPopularFilms,
  urlTopRatedFilms,
  urlComingSoonFilms,
  urlByGenreFilms,
  urlBySearchFilms,
} from '../utils/constants';

export const FETCH_FILMS_REQUEST = 'FETCH_FILMS_REQUEST';
export const FETCH_FILMS_SUCCESS = 'FETCH_FILMS_SUCCESS';
export const FETCH_FILMS_ERROR = 'FETCH_FILMS_ERROR';

export const receiveFilmsSuccess = payload => ({
  type: FETCH_FILMS_SUCCESS,
  payload,
});

export const receiveFilmsRequest = () => ({
  type: FETCH_FILMS_REQUEST,
});

export const receiveFilmsError = payload => ({
  type: FETCH_FILMS_ERROR,
  payload,
});

export function fetchFilms(url, filmId, page = 1) {
  return (dispatch) => {
    dispatch(receiveFilmsRequest());
    return axios.get(`${url}&page=${page}`)
      .then((response) => {
        if (filmId) {
          dispatch(receiveFilmsSuccess({ films: response.data.results, url, page }));
          dispatch(receiveMainFilmInfo(filmId));
        } else {
          dispatch(receiveFilmsSuccess({ films: response.data.results, url, page }));
          dispatch(receiveMainFilmInfo(response.data.results[0].id));
        }
      })
      .catch(error => dispatch(receiveFilmsError(error.message)));
  };
}

export function fetchFilmsPopular() {
  return (dispatch) => {
    dispatch(fetchFilms(urlPopularFilms));
  };
}

export function fetchFilmsTopRated() {
  return (dispatch) => {
    dispatch(fetchFilms(urlTopRatedFilms));
  };
}

export function fetchFilmsComingSoon() {
  return (dispatch) => {
    dispatch(fetchFilms(urlComingSoonFilms));
  };
}

export function fetchFilmsByGenre(genreID, filmId) {
  return (dispatch) => {
    dispatch(fetchFilms(`${urlByGenreFilms}&with_genres=${genreID}`, filmId));
  };
}

export function fetchFilmsBySearch(word, filmId) {
  return (dispatch) => {
    dispatch(fetchFilms(`${urlBySearchFilms}&query=${word}`, filmId));
  };
}

export function fetchNextFilms(filmId) {
  return (dispatch, getState) => {
    const state = getState();
    if (state.films.lastPage !== -1) {
      dispatch(fetchFilms(
        state.films.url,
        filmId,
        state.films.lastPage + 1,
      ));
    }
  };
}
