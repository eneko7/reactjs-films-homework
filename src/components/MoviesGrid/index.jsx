import MoviesGrid from './MoviesGridContainer';

export default MoviesGrid;
