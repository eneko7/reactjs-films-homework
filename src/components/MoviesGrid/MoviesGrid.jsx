import React from 'react';
import PropTypes from 'prop-types';
import style from './MoviesGrid.scss';
import MovieElement from './MovieElement';
import MoviesCategories from './MoviesCategories';

class MoviesGrid extends React.Component {
  constructor(props) {
    super(props);
    this.onScrollHandler = this.onScrollHandler.bind(this);
  }

  componentDidMount() {
    global.document.addEventListener('scroll', this.onScrollHandler);
  }

  componentWillUnmount() {
    global.document.removeEventListener('scroll', this.onScrollHandler);
  }

  onScrollHandler() {
    const { fetchNextFilms, isFetchingFilms } = this.props;
    if ((global.window.innerHeight + global.window.pageYOffset)
    >= global.document.body.offsetHeight && !isFetchingFilms) {
      fetchNextFilms();
    }
  }

  render() {
    const { films, isFetchingFilms, genres } = this.props;
    if (films.length === 0) {
      return (
        <main className={style.moviesGrid}>
          <div className={style.moviesGrid_wrapper}>
            <ul className={style.moviesGrid_wrapper_MovieElement_ul}>
              <div className={style.error_search}>
                Unable to find anythig on your request, please change your request and try again...
                <br />
                <br />
                or You can choose one of categories...
              </div>
            </ul>
          </div>
          <MoviesCategories />
        </main>
      );
    }
    const filmsItems = films.map(elem => (
      <li key={elem.id} className={style.moviesGrid_wrapper_MovieElement_ul_item}>
        <MovieElement film={elem} genresList={genres} />
      </li>
    ));
    return (
      <main className={style.moviesGrid}>
        <MoviesCategories />
        <div className={style.moviesGrid_wrapper}>
          <ul className={style.moviesGrid_wrapper_MovieElement_ul}>
            {filmsItems}
          </ul>
          {isFetchingFilms
            && (
              <div>
                <div className={style.moviesGrid_wrapper_loading} id="loading">
                  <div className={style.moviesGrid_wrapper_loading} />
                  <div className={style.moviesGrid_wrapper_loading} />
                  <div className={style.moviesGrid_wrapper_loading} />
                </div>
                <div className={style.moviesGrid_wrapper_loading_text} id="loading_text">
                  <span className={style.moviesGrid_wrapper_loading_text_span}>LOADING</span>
                </div>
              </div>
            )
          }
        </div>
      </main>
    );
  }
}

MoviesGrid.propTypes = {
  fetchNextFilms: PropTypes.func.isRequired,
  isFetchingFilms: PropTypes.bool.isRequired,
  films: PropTypes.arrayOf(PropTypes.object).isRequired,
  genres: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default MoviesGrid;
