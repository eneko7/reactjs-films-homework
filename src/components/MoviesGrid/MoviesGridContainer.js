import { connect } from 'react-redux';
import MoviesGrid from './MoviesGrid';
import { fetchFilmsPopular, fetchNextFilms } from '../../modules/films/filmsActions';
import { getFilms, getIsFetchingFilms } from '../../modules/films/filmsSelectors';

const mapStateToProps = state => ({
  films: getFilms(state),
  isFetchingFilms: getIsFetchingFilms(state),
});

const mapDispatchToProps = {
  fetchFilmsPopular,
  fetchNextFilms,
};
export default connect(mapStateToProps, mapDispatchToProps)(MoviesGrid);
