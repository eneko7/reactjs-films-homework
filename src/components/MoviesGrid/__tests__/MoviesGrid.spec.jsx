import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
// import moxios from 'moxios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import MoviesGrid from '../index';
import mockFilms from '../../../modules/mocks/getFilmsMock';
import mockGenres from '../../../modules/mocks/getGenresMock';
import * as filmsActions from '../../../modules/films/filmsActions';

jest.mock('../Genres', () => () => 'Genres');
jest.mock('../MoviesCategories', () => () => 'MoviesCategories');
const shallow = new ShallowRenderer();
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  films: {
    allFilms: mockFilms.results,
    isFetchingFilms: false,
  },
  genres: {
    allGenres: mockGenres.genres,
    lastGenreID: 0,
  },
  navlinks: {
    clickedLink: 'Trending',
  },
  film: {
    filmTrailer: 'trailer',
    isFetchingFilm: false,
    errorFilm: false,
  },
});
describe('MoviesList Snapshot', () => {
  test('snapshot', () => {
    const component = shallow.render(
      <Provider store={store}>
        <MemoryRouter>
          <MoviesGrid fetchFilmsByGenre={() => ('Hello')} fetchFilmsBySearch={() => ('hello')} />
        </MemoryRouter>
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});
jest.mock('../../../modules/films/filmsActions',
  () => ({
    fetchNextFilms: jest.fn(() => ({ type: 'FETCH_NEXT_FILM' })),
    fetchGenres: jest.fn(() => ({ type: 'FETCH_GENRES_REQUEST' })),
    fetchFilmsPopular: jest.fn(() => ({ type: 'FETCH_FILMS_REQUEST' })),
  }));
jest.mock('../../../modules/navlinks/navlinksActions',
  () => ({
    pushNavigationLink: jest.fn(() => ({ type: 'PUSH_NAVIGATION_LINK' })),
  }));

describe('MoviesGrid logics', () => {
  afterEach(() => {
    filmsActions.fetchNextFilms.mockClear();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });
  // test('dispatch genres and films loading after componentDidMount', (done) => {
  //   TestRenderer.create(
  //     <Provider store={store}>
  //       <MemoryRouter>
  //         <MoviesGrid fetchFilmsByGenre={() => ('Hello')} fetchFilmsBySearch={() => ('hello')} />
  //       </MemoryRouter>
  //     </Provider>,
  //   );
  //   moxios.stubRequest(/api.themoviedb.org/, {
  //     status: 200,
  //     response: { genres: mockGenres },
  //   });
  //   const expectedActions = [
  //     { type: 'PUSH_NAVIGATION_LINK' },
  //   ];

  //   moxios.wait(() => {
  //     expect(store.getActions()).toEqual(expectedActions);
  //     done();
  //   });
  // });
  test('function for fetch next films should be called after scrolling', () => {
    const eventMap = {
      scroll: null,
    };
    global.document.addEventListener = jest.fn((event, cb) => {
      eventMap[event] = cb;
    });
    Object.defineProperty(global.window, 'innerHeight', {
      writable: true,
      value: 1000,
    });
    Object.defineProperty(global.window, 'pageYOffset', {
      writable: true,
      value: 1000,
    });
    Object.defineProperty(global.document.body, 'offsetHeight', {
      writable: true,
      value: 1,
    });
    TestRenderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <MoviesGrid fetchFilmsByGenre={() => ('Hello')} fetchFilmsBySearch={() => ('hello')} />
        </MemoryRouter>
      </Provider>,
    );
    eventMap.scroll();
    expect(filmsActions.fetchNextFilms).toBeCalled();
  });
  test("function for fetch next films shouldn't be called when films are fetching", () => {
    const store2 = mockStore({
      films: {
        allFilms: mockFilms.results,
        isFetchingFilms: true,
      },
      genres: {
        allGenres: mockGenres.genres,
        lastGenreID: 0,
      },
      film: {
        filmTrailer: 'trailer',
        isFetchingFilm: true,
        errorFilm: false,
      },
      navlinks: {
        clickedLink: 'Trending',
      },
    });
    const eventMap = {
      scroll: null,
    };
    global.document.addEventListener = jest.fn((event, cb) => {
      eventMap[event] = cb;
    });
    Object.defineProperty(global.window, 'innerHeight', {
      writable: true,
      value: 1000,
    });
    Object.defineProperty(global.window, 'pageYOffset', {
      writable: true,
      value: 1000,
    });
    Object.defineProperty(global.document.body, 'offsetHeight', {
      writable: true,
      value: 1,
    });
    TestRenderer.create(
      <Provider store={store2}>
        <MemoryRouter>
          <MoviesGrid fetchFilmsByGenre={() => ('Hello')} fetchFilmsBySearch={() => ('hello')} />
        </MemoryRouter>
      </Provider>,
    );
    eventMap.scroll();
    expect(filmsActions.fetchNextFilms).not.toBeCalled();
  });
  test("function for fetch next films shouldn't be called when films are fetching", () => {
    const store3 = mockStore({
      films: {
        allFilms: [],
        isFetchingFilms: true,
      },
      genres: {
        allGenres: mockGenres.genres,
        lastGenreID: 0,
      },
      film: {
        filmTrailer: 'trailer',
        isFetchingFilm: true,
        errorFilm: false,
      },
      navlinks: {
        clickedLink: 'Trending',
      },
    });
    const eventMap = {
      scroll: null,
    };
    global.document.addEventListener = jest.fn((event, cb) => {
      eventMap[event] = cb;
    });
    Object.defineProperty(global.window, 'innerHeight', {
      writable: true,
      value: 1000,
    });
    Object.defineProperty(global.window, 'pageYOffset', {
      writable: true,
      value: 1000,
    });
    Object.defineProperty(global.document.body, 'offsetHeight', {
      writable: true,
      value: 1,
    });
    TestRenderer.create(
      <Provider store={store3}>
        <MemoryRouter>
          <MoviesGrid fetchFilmsByGenre={() => ('Hello')} fetchFilmsBySearch={() => ('hello')} />
        </MemoryRouter>
      </Provider>,
    );
    eventMap.scroll();
    expect(filmsActions.fetchNextFilms).not.toBeCalled();
  });
  test('lifecycle method should have been called', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <MoviesGrid fetchFilmsByGenre={() => ('Hello')} fetchFilmsBySearch={() => ('hello')} />
        </MemoryRouter>
      </Provider>,
    );
    component.unmount();
  });
});
