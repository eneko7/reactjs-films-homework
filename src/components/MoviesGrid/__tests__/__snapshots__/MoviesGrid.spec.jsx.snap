// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList Snapshot snapshot 1`] = `
<MemoryRouter>
  <withRouter(Connect(MoviesGrid))
    fetchFilmsByGenre={[Function]}
    fetchFilmsBySearch={[Function]}
  />
</MemoryRouter>
`;
